raft-tech/TANF-app
# 1610.1 parser error generator

## Summary of Changes
As part of #1610 we need the parser to create `ParserError` objects instead of returning errors as strings. This will connect the file upload to the error report download done in #1122.

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
1. Upload a datafile from the test folder `tdrs-backend/tdpservice/parsers/test/data`
   * Datafiles containing errors should create `ParserError` objects visible from the django admin. Refer to the unit tests to see the parser errors belonging to each file

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Parser generates `ParserError` objects for errors in TANF and SSP section 1 submissions (`T1`, `T2`, `T3`, `M1`, `M2`, and `M3` and schemas)
    * Category 1 (preparsing) errors are generated for missing/multiple headers, multiple trailers, and schema preparsing validators
    * Category 2 (field value) errors are generated for missing required fields and schema field validators. ParserErrors include record `case_number`, `rpt_month_year`, as well as appropriate `content_type` and `object_id`.
+ [x] Submission history on the frontend contains an error report if errors exist for a datafile submission (says "Currently unavailable otherwise, including while parsing is still executing). Error report link contains `.xlsx` extension.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
